### ISSUE TITLE
Make opening `/dev` symlinks in a race-free way easy

### ISSUE THREAD

**Issue by DemiMarie** [systemd/systemd#23179]
**Is your feature request related to a problem? Please describe.**
Opening `/dev` symlinks to block devices is racy

**Describe the solution you'd like**
The symlinks should point to a `/dev/disk/by-diskseq/$DISKSEQ` or `/dev/disk/by-diskseq/${DISKSEQ}p${PARTNUM}` path.  This allows fairly easy checking if the diskseq is correct by checking the final component of the symlink.

**Describe alternatives you've considered**
1. `/dev/disk/by-diskseq/$DISKSEQ` and `/dev/disk/by-diskseq/${DISKSEQ}p${PARTNUM}` are symlinks to the device nodes.  This is easy to implement, but still requires a utility function to open the path in a race-free way.
2. `/dev/disk/by-diskseq` is a mount point, and opening a file under it opens the correct device node atomically.  This is far easier to use, but somewhat tricky to implement.

**The systemd version you checked that didn't have the feature you are asking for**
git main

**Comment by poettering** [comment]
Not sure I follow? You are suggesting to make the kernel names symlinks to the by-diskseq nodes rather than the by-diskseq names symlinks to the kernel names nodes?

That wouldn't help I fear. device nodes in the fs aren't any better in fixating specific devices than symlinks are — conceptually they are just symlinks by another name, where the target is a major/minor pair rather than an fs path. And that major/minor pair actually identifies the device node in kernel – but is aggressively recycled, hence racy to use.

> /dev/disk/by-diskseq is a mount point, and opening a file under it opens the correct device node atomically. This is far easier to use, but somewhat tricky to implement.

You mean, you want to bind mount the device node? Same issue. 

> 1. `/dev/disk/by-diskseq/$DISKSEQ` and `/dev/disk/by-diskseq/${DISKSEQ}p${PARTNUM}` are symlinks to the device nodes.  This is easy to implement, but still requires a utility function to open the path in a race-free way.

This is already implemented in git main.

**Comment by poettering** [comment]
So, I think we shoudl close this, there's nothing we could do about userspace here. Only way I see is if kernel would natively get support for some new fs or so where inodes are exposed that are the device nodes, but where the kernel would ignore the major/minor pair, and instead go by inode nr or so, which one could use as much better identifiers, since 64bit wide and thus something one can avoid to recycle.

**Comment by DemiMarie** [comment]
> Not sure I follow? You are suggesting to make the kernel names symlinks to the by-diskseq nodes rather than the by-diskseq names symlinks to the kernel names nodes?
> 
> That wouldn't help I fear. device nodes in the fs aren't any better in fixating specific devices than symlinks are — conceptually they are just symlinks by another name, where the target is a major/minor pair rather than an fs path. And that major/minor pair actually identifies the device node in kernel – but is aggressively recycled, hence racy to use.

I want to have all of the userspace-generated names (the ones managed by udev) be symlinks to `/dev/disk/by-diskseq/something` and then to use a proper filesystem to fix the remaining races.

> > /dev/disk/by-diskseq is a mount point, and opening a file under it opens the correct device node atomically. This is far easier to use, but somewhat tricky to implement.
> 
> You mean, you want to bind mount the device node? Same issue.

I am referring to mounting a filesystem (either in-kernel or FUSE) at `/dev/disk/by-diskseq`.  That filesystem would, when given a path, look it up by diskseq in an internal cache and grab a reference to the major/minor device number.  That reference would not be released until the inode is purged from the kernel’s cache, and you can’t open an inode that is _not_ in cache.  So this is 100% race-free.

**Comment by poettering** [comment]
Well, that would really require kernel work first, no? i.e. such an fs must be a kernel thing. i.e. convincing greg that there actually is an issue here.

**Comment by DemiMarie** [comment]
> Well, that would really require kernel work first, no? i.e. such an fs must be a kernel thing. i.e. convincing greg that there actually is an issue here.

It could be implemented in user-space using FUSE, sysfs, and the existing uevent netlink APIs.  A kernel implementation is obviously to be preferred, though.  I would start working on it if @marmarek did not have other things I need to be doing, but he does.

**Comment by poettering** [comment]
nah, fuse is a hack. if this is really desirable, then i am sure this should be implemented in the kernel...
